src/network/network.module.ts

```diff
@@ -1,11 +1,213 @@
+    try {
+      const ipDecimalDottedString = options?.ip?.split("/")?.[0] || "192.168.0.0";
+      const maskBinaryNotation = parseInt(options?.ip?.split("/")?.[1] || "24");
+      const maskPrefix = options?.mask ? IPv4Mask.fromDecimalDottedString(options.mask).prefix : maskBinaryNotation;
+      const ipRange = IPv4CidrRange.fromCidr(`${IPv4.fromString(ipDecimalDottedString)}/${maskPrefix}`);
+      const ip = ipRange.getFirst();
+      const mask = ipRange.getPrefix().toMask();
+      const gateway = options?.gateway ? new IPv4(options.gateway) : undefined;
+      const network = await this.deps.networkApi.createNetwork({
+        id: options?.id,
+        ip: ip.toString(),
+        mask: mask?.toString(),
+        gateway: gateway?.toString(),
+      });
+      // add Requestor as network node
+      const requestorId = await this.deps.networkApi.getIdentity();
+      await this.createNetworkNode(network, requestorId, options?.ownerIp);
+      this.deps.logger.info(`Network created`, network.getNetworkInfo());
+      return network;
+    } catch (error) {
+      if (error instanceof GolemNetworkError) {
+        throw error;
+      }
+      throw new GolemNetworkError(
+        `Unable to create network. ${error?.response?.data?.message || error}`,
+        NetworkErrorCode.NetworkCreationFailed,
+        undefined,
+        error,
+      );
+    }
+  }
+  async removeNetwork(network: Network): Promise<void> {
+    try {
+      await this.deps.networkApi.removeNetwork(network);
+      this.deps.logger.info(`Network removed`, network.getNetworkInfo());
+    } catch (error) {
+      throw new GolemNetworkError(
+        `Unable to remove network. ${error}`,
+        NetworkErrorCode.NetworkRemovalFailed,
+        undefined,
+        error,
+      );
+    }
+  }
+  async createNetworkNode(network: Network, nodeId: string, nodeIp?: string): Promise<NetworkNode> {
+    try {
+      if (!network.isNodeIdUnique(nodeId)) {
+        throw new GolemNetworkError(
+          `Network ID '${nodeId}' has already been assigned in this network.`,
+          NetworkErrorCode.AddressAlreadyAssigned,
+          network.getNetworkInfo(),
+        );
+      }
+      let ipv4: IPv4;
+      if (nodeIp) {
+        ipv4 = IPv4.fromString(nodeIp);
+        if (!network.isIpInNetwork(ipv4)) {
+          throw new GolemNetworkError(
+            `The given IP ('${nodeIp}') address must belong to the network ('${network.getNetworkInfo().ip}').`,
+            NetworkErrorCode.AddressOutOfRange,
+            network.getNetworkInfo(),
+          );
+        }
+        if (!network.isNodeIpUnique(ipv4)) {
+          throw new GolemNetworkError(
+            `IP '${nodeIp.toString()}' has already been assigned in this network.`,
+            NetworkErrorCode.AddressAlreadyAssigned,
+            network.getNetworkInfo(),
+          );
+        }
+      } else {
+        ipv4 = network.getFirstAvailableIpAddress();
+      }
```

Comment:
This is really hard to read with all that nesting, what do you think about a refactor to move the logic of finding a free ip to a separate function?
```ts
private getFreeIpInNetwork(network: Network, targetIp?: string): IPv4 {
    if (!targetIp) {
      return network.getFirstAvailableIpAddress();
    }
    const ipv4 = IPv4.fromString(targetIp);
    if (!network.isIpInNetwork(ipv4)) {
      throw new GolemNetworkError(
        `The given IP ('${targetIp}') address must belong to the network ('${network.getNetworkInfo().ip}').`,
        NetworkErrorCode.AddressOutOfRange,
        network.getNetworkInfo(),
      );
    }
    if (!network.isNodeIpUnique(ipv4)) {
      throw new GolemNetworkError(
        `IP '${targetIp.toString()}' has already been assigned in this network.`,
        NetworkErrorCode.AddressAlreadyAssigned,
        network.getNetworkInfo(),
      );
    }
    return ipv4;
  }

  async createNetworkNode(network: Network, nodeId: string, nodeIp?: string): Promise<NetworkNode> {
    if (!network.isNodeIdUnique(nodeId)) {
      throw new GolemNetworkError(
        `Network ID '${nodeId}' has already been assigned in this network.`,
        NetworkErrorCode.AddressAlreadyAssigned,
        network.getNetworkInfo(),
      );
    }
    try {
      const ipv4 = this.getFreeIpInNetwork(network, nodeIp);
      const node = await this.deps.networkApi.createNetworkNode(network, nodeId, ipv4.toString());
      network.addNode(node);
      this.deps.logger.info(`Node has been added to the network.`, { id: nodeId, ip: ipv4.toString() });
      return node;
      //.. rest
```